### DISCUSSION TITLE
bindings also for wasmJS target

### DISCUSSION THREAD

**Discussion by sergeych** [ionspin/kotlin-multiplatform-libsodium#44] [Ideas]

**Comment by ionspin** [comment]
Hi @sergeych,

this would be nice to add, but I currently don't exactly know where to start with this. Do you have any pointers?

**Comment by sergeych** [comment]
Hi! I'm as new in new kotlin wamsJS as anybody else, though I would try two aaproaches:

First of all I'm sorry to say I didn't get deep into your nice library guts, so probably things could be easier than I think.

1. Fast and dirty hack. It is possible to use [wasmJS-JS interoperability](https://kotlinlang.org/docs/wasm-js-interop.html#use-kotlin-code-in-javascript) to call your existing JS package (maybe providing some extra JS-friendly endpoints for it). So JS target is compiled first then it's compiled code is used in wasmJS. Though dirty hack, it might work really well and cost less than any other approaches.

2. libsodium itself could be compiled to wasm, probably you already use it in this project? Anyway, the compiled wasm could be called via JS interop (1). In the future it even could be possible to merge compiled libsodium into the same wasm, although I'm sure right now it is too early for it (wasmJS target will most likely change too much so the efforts could be in vane soon).

Actually, at this time (2) is not much better solution than (1), not until we can merge wasm libraries / rust/C code into wasm JS target.

**Comment by ionspin** [comment]
Those are interesting directions, and you are right, the libsodiumJS library that is used by JS target is wasm produced by emscripten (https://github.com/jedisct1/libsodium.js/), so wasm is already used. Let's convert this into a discussion and keep it open to track the developments regarding wasmJS.

**Comment by sergeych** [comment]
@ionspin So, I have invested some time and money and hired 2 young talents from MTU graduates to create KMP wasmJS variant based on wxisting JS/wasm code of yours: https://github.com/sergeych/kotlin-multiplatform-libsodium @ionspin could you please check this? while ugly (your build system was a challenge for modern kotlin), it provides absolutely working solution that I do use in several my work products for few month with Compose/wasmJS very well. I think you could adopt it: it is way too big a change inthe build system, as for me, to create a PR, but it works pretty well.

**Comment by sergeych** [comment]
So, is wasmJs already supported in your newer releases?
